### ISSUE TITLE
Incorrect counting of primary pages in gstat

### ISSUE THREAD

**Issue by ilya071294** [FirebirdSQL/firebird#7624]
FB3/4/5 are affected.
How to reproduce:
1. Prepare the database (page size 4096, record size is about 1500):
```
create database '/tmp/empty.fdb' page_size 4096;

create table TEST(str char(1500));
commit;

set term ^;
execute block as
    declare variable i integer;
begin
    i = 18;
    while (i > 0) do
    begin
        insert into TEST values ('012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789');
        i = i - 1;
    end
end^

set term ;^
commit;
```
2. Run gstat with -d option.
```
gstat /tmp/empty.fdb -d
...
Analyzing database pages ...
TEST (128)
    Primary pointer page: 284, Index root page: 285
    Pointer pages: 1, data page slots: 16
    Data pages: 16, average fill: 42%
    Primary pages: 16, secondary pages: 0, swept pages: 0
    Empty pages: 7, full pages: 8
    Fill distribution:
     0 - 19% = 7
    20 - 39% = 0
    40 - 59% = 0
    60 - 79% = 9
    80 - 99% = 0
```
9 primary pages are expected here (18 records, 2 per page) but we get 16. Looks like empty pages are also included in this count.

**Comment by ilya071294** [comment]
I can prepare the PR for master. Should I fix it in FB3/FB4?

**Comment by hvlad** [comment]
Yes, empty pages are allocated as primary. What is the problem ?

**Comment by ilya071294** [comment]
> Yes, empty pages are allocated as primary. What is the problem ?

In this case the first page from an extent is primary, the last 7 pages are empty (and they may become secondary if it's needed). It's wrong to include empty pages in primary pages. Also primary + secondary + empty should not be greater than "data pages" or "data page slots".

**Comment by hvlad** [comment]
Why empty page could not be considered as primary (or secondary) also ?
This is not mutually exclusive flags, this is your mistake.
What is a real problem ?

**Comment by ilya071294** [comment]
> Why empty page could not be considered as primary (or secondary) also ?

Because an empty page is neither primary nor secondary.

> What is a real problem ?

The problem is gstat doesn't show the actual information. Primary pages are pages that contain primary records. This database has only 9 data pages with primary records.

**Comment by hvlad** [comment]
Again, you wrongly interpret flags meanings. Primary\secondary and empty\full is ortogonal flags. 
Don't add empty to the primary+secondary (as you don't add full or swept here).

**Comment by ilya071294** [comment]
> Again, you wrongly interpret flags meanings. Primary\secondary and empty\full is ortogonal flags.

I know that empty pages without dpg_secondary flag are counted as primary, and empty pages with dpg_secondary are counted as secondary. But for me it doesn't make sense because:
1. If I update all records and rollback the transaction, I get:
```
Analyzing database pages ...
TEST (131)
    Primary pointer page: 285, Index root page: 286
    Pointer pages: 1, data page slots: 16
    Data pages: 16, average fill: 42%
    Primary pages: 9, secondary pages: 7, swept pages: 0
    Empty pages: 7, full pages: 9
    Fill distribution:
         0 - 19% = 7
        20 - 39% = 0
        40 - 59% = 0
        60 - 79% = 9
        80 - 99% = 0
```
These 7 empty pages became secondary on update and then became empty again on rollback. But now they are counted as secondary. As a developer I understand this. But from a user perspective it may be confusing like "why primary page count changed if I just updated records and didn't delete anything?".
2. If we look at how the engine works with data pages, it's clear that empty pages are not considered either primary or secondary. They are just empty and ready to become primary or secondary depending on the next action. The ppg_dp_secondary/dpg_secondary flag matters only when the data page is not empty. And I don't understand why the logic in gstat cannot be the same.

**Comment by aafemt** [comment]
From user POV it is logical that data pages = primary pages + secondary pages. Changing it would raise more questions than solve. Also not counting empty pages at all would raise "where my DB missing pages are" question.

**Comment by ilya071294** [comment]
> From user POV it is logical that data pages = primary pages + secondary pages. Changing it would raise more questions than solve. Also not counting empty pages at all would raise "where my DB missing pages are" question.

Currently "Data pages" is correct. 9 primary + 0 secondary + 7 empty = 16. What do you mean by "not counting empty pages at all"? gstat also shows "Empty pages" separately.

**Comment by romansimakov** [comment]
> From user POV it is logical that data pages = primary pages + secondary pages. Changing it would raise more questions than solve. Also not counting empty pages at all would raise "where my DB missing pages are" question.

For me it's logical that data pagse = primary page + secondary pages + empty pages.
Empty pages are part of the pages reserved for a relation. I see your and Vlad's points but isn't it to complex for user to use 2 set of 2 flags? A page may be empty, primary or secondary. It's much more simple (and clear) than "A page may be primary, secondary, empty and primary (but convertable to secondary), empty and secondaty (but convertable to primary)".

**Comment by hvlad** [comment]
> For me it's logical that data pages = primary pages + secondary pages + empty pages.

Why you ever need to sum primary, secondary and empty pages ? For what purpose ?
For consistency it is more that enough to know that:
```
data pages = primary + secondary
empty <= data pages
full <= data pages
empty + full <= data pages
swept <= primary
```
Something is logical for you, something else is logical for me, something very else for them...
So far no user get problem with current output.
And some of them have own parsers.
And there is no valuable reason (by my logic) to change things just because someone like that and not this.

**Comment by romansimakov** [comment]
Why you ever need to know number of primary pages?:)
(especially empties)

**Comment by hvlad** [comment]
Ok, you both almost convinced me. 
But ticket should be renamed like "Improve gstat output to not include empty pages into primary\secondary", imho.
